File: src/CorporateMessengerWithAgile/Domain/Result/Result.cs
using System.Runtime.CompilerServices;
using System.Text.Json.Serialization;

namespace Domain.Result
{
    [JsonConverter(typeof(ResultConverterFactory))]
    public class Result
    {
        public bool IsSuccess { get; init; }
        [JsonIgnore]
        public bool IsFailure => !IsSuccess;

        private readonly Error? _error;

        public Error Error
        {
            get
            {
                if (IsSuccess) throw new Exception("Result.CannotAccessExceptionOnSuccess. Не возможно получить доступ к свойству Exception при успешном результате.");
                return _error!;
            }
        }

        public int StatusCode { get; protected set; }

        protected internal Result(bool isSuccess, Error error, int statusCode)
        {
            IsSuccess = isSuccess;
            _error = error;
            StatusCode = statusCode;
        }

        public static Result Success(int statusCode) => new(true, null!, statusCode);
        public static Result Failure(Error error) => new(false, error, error.StatusCode);
        public static Result<T> Success<T>(T value, int statusCode) => new(true, null!, statusCode, value);
        public static Result<T> Failure<T>(Error error) => new(false, error, error.StatusCode, default!);

        public static implicit operator Result(Error error) => Failure(error);

        public Result SetStatusCode(int statusCode)
        {
            StatusCode = statusCode;
            return this;
        }
    }
}

File: src/CorporateMessengerWithAgile/ReactApp/reactapp.client/src/models/result/Result.ts
import { AppError } from './AppError';

export class Result<T = void> {
    private readonly _isSuccess: boolean;
    private readonly _error?: AppError;
    private readonly _value?: T;

    protected constructor(isSuccess: boolean, error?: AppError, value?: T) {
        this._isSuccess = isSuccess;
        this._error = error;
        this._value = value;
    }

    get isSuccess(): boolean { return this._isSuccess; }
    get isFailure(): boolean { return !this._isSuccess; }

    get error(): AppError {
        if (this._isSuccess) throw new Error('Result.CannotAccessExceptionOnSuccess. Не возможно получить доступ к свойству error при успешном результате.');
        return this._error!;
    }
    get value(): T {
        if (!this._isSuccess) throw new Error('Result.CannotAccessValueOnFailure. Не возможно получить доступ к свойству value для неудачного результата.');
        return this._value!;
    }

    static Success(): Result { return new Result(true, undefined, undefined); }
    static Failure(error: AppError): Result { return new Result(false, error, undefined); }
    static SuccessWith<T>(value: T): Result<T> { return new Result(true, undefined, value); }
    static FailureWith<T>(error: AppError): Result<T> { return new Result(false, error, undefined as T); }
}